PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp

```diff
@@ -102,15 +85,32 @@ HTREEITEM TreeView::addItem(const TCHAR *itemName, HTREEITEM hParentItem, int iI
 	tvi.iImage = iImage;//isNode?INDEX_CLOSED_NODE:INDEX_LEAF; 
 	tvi.iSelectedImage = iImage;//isNode?INDEX_OPEN_NODE:INDEX_LEAF; 
 
-	// Save the full path of file in the item's application-defined data area. 
-	tvi.lParam = (filePath == NULL?0:(LPARAM)(new generic_string(filePath)));
+	// Save the full path of file in the item's application-defined data area.
+
+	tvi.lParam = (LPARAM)(data);
 
 	TVINSERTSTRUCT tvInsertStruct;
 	tvInsertStruct.item = tvi; 
-	tvInsertStruct.hInsertAfter = (HTREEITEM)TVI_LAST;
+	tvInsertStruct.hInsertAfter = TVI_LAST;
 	tvInsertStruct.hParent = hParentItem;
 
-	return (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	if (insertFunc)
+	{
+		TVITEM tvItem;
+		tvItem.mask = TVIF_PARAM;
+		tvItem.hItem = hParentItem;
+		SendMessage(_hSelf, TVM_GETITEM, 0,(LPARAM)&tvItem);
+		TreeViewData* parentData = (TreeViewData*) tvItem.lParam;
+		tvInsertStruct.hInsertAfter = insertFunc(this, hParentItem, parentData, itemName, data);
+	}
+
+	HTREEITEM newItem = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	_validHandles.insert(newItem);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.